src/debug.rs

```diff
@@ -0,0 +1,215 @@
+//!
+//! This module is still present when the `log` feature is disabled, but all
+//! logging operations are redacted. Redaction completely compiles out log
+//! statements: not even the format strings remain in the final binary.
+//!
+//! Manticore code *should not* call into the [`log`] crate directly outside of
+//! this module.
+
+// TODO: Remove this once we start using these macros in Manticore.
+#![allow(unused)]
+
+use core::fmt;
+
+#[cfg(doc)]
+use __raw_log as log;
+
+/// A wrapped Manticore error.
+///
+/// This type should always be referred to as `manticore::Error`. It represents
+/// an error with extra (potentially redacted) information attached. This type
+/// cannot be directly created by users of the library.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Error<E> {
+    inner: E,
+}
+
+impl<E> Error<E> {
+    /// Creates a new `Error`. This function is an implementation detail,
+    /// and should not be called by users.
+    #[doc(hidden)]
+    pub fn __new(inner: E) -> Self {
+        Self { inner }
+    }
+
+    /// Transforms the wrapper error by way of an [`Into`] conversion.
+    ///
+    /// Generally, this function should not be necessary, because Manticore-
+    /// defined error types manually implement the relevant [`From`]
+    /// implementations for `manticore::Error`, which in turn call `cast()`.
+    pub fn cast<F: From<E>>(self) -> Error<F> {
+        Error {
+            inner: self.inner.into(),
+        }
+    }
+
+    /// Gets the wrapped error.
+    pub fn into_inner(self) -> E {
+        self.inner
+    }
+}
+
+impl<E> AsRef<E> for Error<E> {
+    fn as_ref(&self) -> &E {
+        &self.inner
+    }
+}
+
+impl<E> AsMut<E> for Error<E> {
+    fn as_mut(&mut self) -> &mut E {
+        &mut self.inner
+    }
+}
+
+impl<E: fmt::Display> fmt::Display for Error<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        struct DisplayAsDebug<'a, E>(&'a E);
+        impl<E: fmt::Display> fmt::Debug for DisplayAsDebug<'_, E> {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                self.0.fmt(f)
+            }
+        }
+
+        f.debug_struct("manticore::Error")
+            .field("inner", &DisplayAsDebug(&self.inner))
+            .finish()
+    }
+}
+
+/// Generates `From` implementations for `manticore::Error`.
+///
+/// We would like to write this `impl`:
+/// ```compile_fail
+/// # use manticore::Error;
+/// impl<E1, E2> From<Error<E1>> for Error<E2> where E2: From<E1> {
+///     fn from(e: Error<E1>) -> Error<E2> {
+///         e.cast()
+///     }
+/// }
+/// ```
+///
+/// Unfortunately, the trait coherence rules mean that for `E1 == E2`,
+/// this conflicts with the standard library's `impl<T> From<T> for T`
+/// impl. We work around this by calling this macro for every Manticore
+/// error definition that has `From` impls.
+macro_rules! debug_from {
+    ($e:ident<$trait:ident: $bound:path> => $($f:ty),+ $(,)?) => {$(
```

Comment:
This first macro rule is an ugly hack so that this macro can be used with `Error<Header>` in src/server/handler.rs. [See usage](https://github.com/lowRISC/manticore/pull/123/files?authenticity_token=G3yXZk0zkJbfp64g1zPlHUGY-1JgnaE_n28D_LkqSyzV6OfxWXs2dp4q6S7mZ--4mIwZGFZVJ4KwtjeJdfQQfg&file-filters%5B%5D=.rs#diff-4c9251cd3e4426cdcd6c86b12087130f3b5a7cf27db998d8dbb648d3de6c7237R126). It works... but I'm not too proud of it. I'm super new to Rust macros, but based on what I've read, there's not really a clean way of doing this besides resorting to proc macros, which seems like a rabbit hole not worth tackling yet. That said, feedback would be appreciated if you think otherwise.

Reply:
Not really. One option to cut down on repetition is to define something like

```
macro_rules! debug_from {
  (@actual $e:ident[$($tparams:tt)*] => ...) => {
    // Use `$($tparams)* as appropriate in <>
  };
  ($e:ident => ...) => {debug_from!(@actual $e[] => ...)};
}
```

Note that wherever a `<>` is allowed, you can leave it empty, e.g. `i32<>` is valid Rust. You can use this to your advantage here.

Reply:
I bashed my head against this a bit, but I'm struggling to get any sort of recursive calls to work as I'd expect. For now, I combined two of the macro arms by matching on the `tt` type.

Reply:
Sg. We can bash on it more later.